### PR TITLE
Issue #457

### PR DESCRIPTION
**Pull request by ljacomet**
* Make ServiceLocator.discoverService private
* Single path for service initialization through ServiceLocator
* Solution for service initialization without configuration
* Improve error reporting when no persistence service is found

### CODE REVIEW

File: 107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java [hunk 1]
```diff
@@ -118,13 +118,14 @@ public CacheManager getCacheManager(URI uri, ClassLoader classLoader, Properties
     return cacheManager;
   }
 
+  @ServiceDependency(services = ManagementRegistry.class)
```
Comment (chrisdennis) [inline 107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java]: This (and other instances) can be done without the `services = ` right?

File: core/src/main/java/org/ehcache/spi/ServiceLocator.java [hunk 2]
```diff
@@ -145,8 +136,17 @@ void addService(final Service service, final boolean expectsAbstractRegistration
         }
       }
 
+      if (service.getClass().isAnnotationPresent(ServiceDependency.class)) {
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/spi/ServiceLocator.java]: Could we just to this with the `getAnnotation` call and a null check?

File: core/src/main/java/org/ehcache/spi/ServiceLocator.java [hunk 3]
```diff
@@ -164,25 +164,28 @@ void addService(final Service service, final boolean expectsAbstractRegistration
...
   }
 
   @Override
-  public <T extends Service> T findService(Class<T> serviceType) {
-    return findService(serviceType, null);
+  public <T extends Service> T getService(Class<T> serviceType) {
+    return findService(serviceType, null, false);
+  }
+
+  public <T extends Service> T getOrCreateService(Class<T> serviceType) {
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/spi/ServiceLocator.java]: Crazy idea... could we eliminate this method entirely by adding services that list the mandatory services as their dependencies?

File: 107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java (hunk 1 above)
Comment (ljacomet) [inline 107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java]: Only by renaming `services()` to `value()` but as we have only one annotation parameter, this seems to be the convention. Will do that.

File: core/src/main/java/org/ehcache/spi/ServiceLocator.java (hunk 2 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/spi/ServiceLocator.java]: Yep, especially given the implementation of `isAnnotationPresent`. Still, I prefer the current style - feels more "correct" and less `null` tied.

File: core/src/main/java/org/ehcache/spi/ServiceLocator.java (hunk 3 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/spi/ServiceLocator.java]: I do not think so. Using `getOrCreateService` still allows the `Service` instance to be resolved by finding a proper `ServiceFactory` or just be a no-op when a `Service` instance has been provided at configuration.

**Comment by ljacomet** [comment]
All comments have been addressed, including the crazy idea ... see added commit for details

**Pull request merged**